onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -32,22 +32,24 @@
 namespace onnxruntime {
 namespace session_state_utils {
 
-static common::Status ReserveUsingArenaFromShapeAndType(const TensorShape& tensor_shape, const DataTypeImpl* type,
-                                                        const AllocatorPtr& arena_alloc, /*out*/ void*& p_data) {
-  if (arena_alloc->Info().alloc_type != OrtArenaAllocator)
-    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Given allocator must be an arena based allocator");
-
+// The following method will allocate memory directly using the device allocator.
+// It can handle arena-based allocators and non-arena based allocators.
+static common::Status AllocateBufferUsingDeviceAllocatorFromShapeAndType(const TensorShape& tensor_shape, const DataTypeImpl* type,
```

Comment:
The method name suggests that that we should always use the device allocator. But we use the arena allocator when alloc_type == OrtArenaAllocator. Now I understand that Reserve internally calls the device allocator, but this implementation is a quirk of Reserve and unless someone goes and looks inside Reserve the usage of arena allocator is confusing in this method from a readability standpoint. Actually I liked the previous method name where it was clear what the function was doing. If not, can we add a clear comment near the Reserve invocation that says this is going to use the device allocator?

Overall, I think the usage of Reserve inside the arena seems like a hack. It also makes the interface for arena allocators a bit non-standard. But that's a separate discussion.

Reply:
The previous method only did a Reserve() and the naming reflected that. Scott preferred that this method did Reserve (for arena based allocators) and Alloc (for non-arena based allocators) so that this method is "more inclusive" in what it handles and its behavior with dealing with different type of allocators is an implementation detail. Now that this method doesn't do just Reserve anymore the naming had to change. 

I can add a comment stating that Reserve actually does a device allocation.